packages-exp/auth-exp/src/core/util/handler.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SDK_VERSION } from '@firebase/app-exp';
+import * as externs from '@firebase/auth-types-exp';
+import { ApiKey, AppName, Auth } from '../../model/auth';
+import { AuthEventType } from '../../model/popup_redirect';
+import { AuthErrorCode } from '../errors';
+import { OAuthProvider } from '../providers/oauth';
+import { _assert } from './assert';
+import { isEmpty, querystring } from '@firebase/util';
+import { _emulatorUrl } from './emulator';
+
+/**
+ * URL for Authentication widget which will initiate the OAuth handshake
+ *
+ * @internal
+ */
+const WIDGET_PATH = '__/auth/handler';
+
+/**
+ * URL for emulated environment
+ *
+ * @internal
+ */
+const EMULATOR_WIDGET_PATH = 'emulator/auth/handler';
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type WidgetParams = {
+  apiKey: ApiKey;
+  appName: AppName;
+  authType: AuthEventType;
+  redirectUrl?: string;
+  v: string;
+  providerId?: string;
+  scopes?: string;
+  customParameters?: string;
+  eventId?: string;
+  tid?: string;
+} & { [key: string]: string | undefined };
+
+export function _getRedirectUrl(
```

Comment:
This is moved from `platform_browser/popup_redirect`. The only change is that it now has an "additionalParams" parameter